modules/photo/src/contrast_preserve.hpp

```diff
@@ -81,350 +83,331 @@ int round_num(double a)
-    gradvector(lab_channel[2],Imb);
+    gradvector(lab_channel[0], ImL);
+    gradvector(lab_channel[1], Ima);
+    gradvector(lab_channel[2], Imb);
 
-    double res =0.0;
-    for(unsigned int i=0;i<ImL.size();i++)
+    Cg.resize(ImL.size());
+    for (unsigned int i = 0; i<ImL.size(); i++)
     {
-        res=sqrt(pow(ImL[i],2) + pow(Ima[i],2) + pow(Imb[i],2))/100;
-        Cg.push_back(res);
+        const double res = sqrt(pow(ImL[i], 2) + pow(Ima[i], 2) + pow(Imb[i], 2)) / 100;
+        Cg[i] = res;
     }
-
-    ImL.clear();
-    Ima.clear();
-    Imb.clear();
 }
 
-void Decolor::add_vector(vector < vector <int> > &comb, int &idx, int r,int g,int b)
+void Decolor::add_vector(vector<Vec3i> &comb, int &idx, int r, int g, int b)
 {
-    comb.push_back( vector <int>() );
-    comb.at(idx).push_back( r );
-    comb.at(idx).push_back( g );
-    comb.at(idx).push_back( b );
+    comb.push_back(Vec3i(r, g, b));
     idx++;
 }
 
-void Decolor::add_to_vector_poly(vector < vector <double> > &polyGrad, vector <double> &curGrad, int &idx1)
+void Decolor::add_to_vector_poly(vector<vector<double> > &polyGrad, const vector<double> &curGrad, int &idx1)
 {
-    polyGrad.push_back( vector <double>() );
-    for(unsigned int i=0;i<curGrad.size();i++)
-        polyGrad.at(idx1).push_back(curGrad[i]);
+    polyGrad.push_back(curGrad);
     idx1++;
 }
 
-void Decolor::weak_order(Mat img, vector <double> &alf)
+void Decolor::weak_order(const Mat &img, vector <double> &alf) const
 {
-    int h = img.size().height;
-    int w = img.size().width;
-    double sizefactor;
-    if((h + w) > 800)
+    const int h = img.size().height;
+    const int w = img.size().width;
+    if ((h + w) > 800)
     {
-        sizefactor = (double)800/(h+w);
-        resize(img,img,Size(round_num(h*sizefactor),round_num(w*sizefactor)));
+        const double sizefactor = 800.0 / (h + w);
+        resize(img, img, Size(round_num(h*sizefactor), round_num(w*sizefactor)));
     }
 
-    Mat curIm = Mat(img.size(),CV_32FC1);
-    vector <Mat> rgb_channel;
-    split(img,rgb_channel);
+    Mat curIm = Mat(img.size(), CV_32FC1);
+    Mat rgb_channel[3];
+    split(img, rgb_channel);
 
     vector <double> Rg, Gg, Bg;
-    vector <double> t1, t2, t3;
-    vector <double> tmp1, tmp2, tmp3;
+    gradvector(rgb_channel[2], Rg);
+    gradvector(rgb_channel[1], Gg);
+    gradvector(rgb_channel[0], Bg);
 
-    gradvector(rgb_channel[2],Rg);
-    gradvector(rgb_channel[1],Gg);
-    gradvector(rgb_channel[0],Bg);
+    vector <double> t1(Rg.size()), t2(Rg.size()), t3(Rg.size());
+    vector <double> tmp1(Rg.size()), tmp2(Rg.size()), tmp3(Rg.size());
 
-    double level = .05;
+    const double level = .05;
 
-    for(unsigned int i=0;i<Rg.size();i++)
+    for (unsigned int i = 0; i<Rg.size(); i++)
     {
-        if(Rg[i] > level)
-            t1.push_back(1.0);
+        if (Rg[i] > level)
+            t1[i] = (1.0);
         else
-            t1.push_back(0.0);
+            t1[i] = (0.0);
```

Comment:
Use this form:
```
t1[i] = (Rg[i] > level) ? 1.0 : 0.0;
```

Reply:
Thanks!  I will try this change in the next refactoring patch.